File: Backend/controllers/authControllers.js
import catchAsyncErrors from '../middleware/catchasyncError.js';
import User from '../models/user.js';
import sendToken from '../utils/sendToken.js';
import {resetPasswordEmailTemplate} from '../utils/emailTemplates.js';
import ErrorHandler from '../utils/errorHandler.js';
import sendEmail from '../utils/sendEmail.js';
import crypto from 'crypto';



//Register a new user
// This function handles user registration by creating a new user in the database
export const registerUser = catchAsyncErrors (async (req, res, next) => {
    const { name, email, password } = req.body;

    const user = await User.create({
        name,email,password
    })

   sendToken(user, 201, res);

});

//Login user
export const loginUser = catchAsyncErrors(async (req, res, next) => {
    const { email, password } = req.body;

    // Check if email and password are provided
    if (!email || !password) {
        return res.status(400).json({ message: 'Please enter email and password' });
    }

    // Find user by email and check password
    const user = await User.findOne({ email }).select('+password');
    if (!user || !(await user.matchPassword(password))) {
        return res.status(401).json({ message: 'Invalid email or password' });

    }

    //check if password matches
    const isPasswordMatch = await user.matchPassword(password);
    
    // If password does not match, return an error
    if (!isPasswordMatch) {
        return res.status(401).json({ message: 'Invalid email or password' });
    }


        
   sendToken(user, 200, res);
});


//Forgot password
export const forgotPassword = catchAsyncErrors(async (req, res, next) => {
    const user = await User.findOne({ email: req.body.email });
    if (!user) {
        return res.status(404).json({ message: 'User not found' });
    }

    // Generate reset token
    const resetToken = user.getResetPasswordToken();
    await user.save();
    const resetUrl = `${process.env.FRONTEND_URL}/api/v1/password/reset/${resetToken}`;

    const message = resetPasswordEmailTemplate(user?.name, resetUrl);
    try{
        
        await sendEmail({
            email: user.email,
            subject: 'Password Recovery',
            message,
        });

        res.status(200).json({
            success: true,
            message: `Email sent to ${user.email} successfully`,
        });

    
    }
    catch (error) {
        user.resetPasswordToken = undefined;
        user.resetPasswordExpire = undefined;
        await user.save();
        return next(new ErrorHandler(error.message, 500));
    }
});


//logout user
export const logoutUser = catchAsyncErrors(async (req, res, next) => {
    res.cookie('token', null, {
        expires: new Date(Date.now()),
        httpOnly: true,
    });

    res.status(200).json({
        success: true,
        message: 'Logged out successfully',
    });
});


//Reset password
export const resetPassword = catchAsyncErrors(async (req, res, next) => {
    // Hash the url token
    const resetPasswordToken = crypto
        .createHash('sha256')
        .update(req.params.token)
        .digest('hex');

    // Find user by the reset token and check if token has not expired
    const user = await User.findOne({
        resetPasswordToken,
        resetPasswordExpire: { $gt: Date.now() },
    });

    if (req.body.password !== req.body.confirmPassword) {
        return next(new ErrorHandler('Password does not match', 400));
    }

    if (!user) {
        return res.status(400).json({ message: 'Invalid or expired password reset token' });
    }

    

    // Set the new password
    user.password = req.body.password;
    user.resetPasswordToken = undefined;
    user.resetPasswordExpire = undefined;

    //confirm password

 

    await user.save();

    sendToken(user, 200, res);
});

// Upload user avatar   =>  /api/v1/me/upload_avatar
export const uploadAvatar = catchAsyncErrors(async (req, res, next) => {
  const avatarResponse = await upload_file(req.body.avatar, "../public/images");

  // Remove previous avatar
  if (req?.user?.avatar?.url) {
    await delete_file(req?.user?.avatar?.public_id);
  }

  const user = await User.findByIdAndUpdate(req?.user?._id, {
    avatar: avatarResponse,
  });

  res.status(200).json({
    user,
  });
});



// Get current logged in user details
export const getCurrentUserDetails = catchAsyncErrors(async (req, res, next) => {
    const user = await User.findById(req.user.id);

    if (!user) {
        return next(new ErrorHandler('User not found', 404));
    }

    res.status(200).json({
        success: true,
        user,
    });
}
);

//change password
export const changePassword = catchAsyncErrors(async (req, res, next) => {
    const user = await User.findById(req.user.id).select('+password');

    //check previous user password
    const isPasswordMatch = await user.matchPassword(req.body.oldPassword);
    if (!isPasswordMatch) {
        return next(new ErrorHandler('Old password is incorrect', 400));
    }

      // Check confirm password
   if (req.body.newPassword != req.body.confirmPassword) {
    return next(new ErrorHandler("Password does not match", 400));
    }

    user.password = req.body.newPassword;
    await user.save();

    sendToken(user, 200, res);
});

//update user profile
export const updateUserProfile = catchAsyncErrors(async (req, res, next) => {
    const newUserData = {
        name: req.body.name,
        email: req.body.email,
    };

    // Update avatar: TODO

    const user = await User.findByIdAndUpdate(req.user.id, newUserData, {
        new: true,
        runValidators: true,
        useFindAndModify: false,
    });

    res.status(200).json({
        success: true,
        user,
    });
});

//get all admin users
export const getAllAdminUsers = catchAsyncErrors(async (req, res, next) => {
    const users = await User.find({ role: 'admin' });

    if (!users) {
        return next(new ErrorHandler('No admin users found', 404));
    }

    res.status(200).json({
        success: true,
        users,
    });
}
);

export const getAdminUserById = catchAsyncErrors(async (req, res, next) => {
    const users = await User.findById(req.user.id, { role: 'admin' });

    if (!users) {
        return next(new ErrorHandler(`User does not exist with id: ${req.params.id}`));
    }

    res.status(200).json({
        success: true,
        users,
    });
}
);
//update user profile -- admin
export const updateUserProfileByAdmin = catchAsyncErrors(async (req, res, next) => {
    const newUserData = {
        name: req.body.name,
        email: req.body.email,
        role: req.body.role,
    };

    // Update avatar: TODO

        const user = await User.findByIdAndUpdate(req.params.id, newUserData, {
            new: true,
            runValidators: true,
            useFindAndModify: false,
        });
    
        res.status(200).json({
            success: true,
            user,
        });
    });


    export const deleteUserByAdmin = catchAsyncErrors(async (req, res, next) => {
        const user = await User.findByIdAndDelete(req.params.id);
    
        if (!user) {
            return next(new ErrorHandler(`User does not exist with id: ${req.params.id}`));
        }
    
        // Remove avatar from cloudinary - TODO
    
       // await user.remove();
    
        res.status(200).json({
            success: true,
            message: 'User deleted successfully',
        });
    });










    

File: Backend/controllers/authControllers.py
import hashlib
import os
from datetime import datetime, timezone
from typing import Optional

from fastapi import Request, Response, UploadFile
from fastapi.responses import JSONResponse

from Backend.models.user import User
from Backend.utils.emailTemplates import reset_password_email_template
from Backend.utils.errorHandler import ErrorHandler
from Backend.utils.sendEmail import send_email
from Backend.utils.sendToken import send_token


# ── Auth ──────────────────────────────────────────────────────────────────────

async def register_user(request: Request, response: Response) -> JSONResponse:
    """Create a new user account and return a JWT cookie.

    Mirrors: registerUser in authControllers.js
    POST /api/v1/register
    """
    data = await request.json()
    name = data.get("name")
    email = data.get("email")
    password = data.get("password")

    existing = await User.find_one(User.email == email)
    if existing:
        raise ErrorHandler("Email already registered", 400)

    user = User(name=name, email=email, password=password)
    await user.insert()

    token_data = send_token(user, 201, response)
    return JSONResponse(
        {"success": True, "user": _serialize_user(user), **token_data},
        status_code=201,
    )


async def login_user(request: Request, response: Response) -> JSONResponse:
    """Validate credentials and return a JWT cookie.

    Mirrors: loginUser in authControllers.js
    POST /api/v1/login
    """
    data = await request.json()
    email = data.get("email")
    password = data.get("password")

    if not email or not password:
        raise ErrorHandler("Please enter email and password", 400)

    # password field has select:false equivalent — fetch it explicitly
    user = await User.find_one(User.email == email)
    if not user:
        raise ErrorHandler("Invalid email or password", 401)

    if not user.match_password(password):
        raise ErrorHandler("Invalid email or password", 401)

    token_data = send_token(user, 200, response)
    return JSONResponse(
        {"success": True, "user": _serialize_user(user), **token_data},
        status_code=200,
    )


async def logout_user(response: Response) -> JSONResponse:
    """Clear the JWT cookie.

    Mirrors: logoutUser in authControllers.js
    POST /api/v1/logout  (note: original was GET, fixed to match frontend)
    """
    response.delete_cookie("token")
    return JSONResponse({"success": True, "message": "Logged out"})


async def forgot_password(request: Request) -> JSONResponse:
    """Send a password-reset email.

    Mirrors: forgotPassword in authControllers.js
    POST /api/v1/password/forgot
    """
    data = await request.json()
    email = data.get("email")

    user = await User.find_one(User.email == email)
    if not user:
        raise ErrorHandler("User not found with this email", 404)

    raw_token = user.get_reset_password_token()
    await user.save()

    frontend_url = os.getenv("FRONTEND_URL", "http://localhost:3000")
    reset_url = f"{frontend_url}/password/reset/{raw_token}"

    message = reset_password_email_template(user.name, reset_url)
    await send_email({"email": user.email, "subject": "EcommerceApp Password Reset", "message": message})

    return JSONResponse({"success": True, "message": f"Email sent to {user.email}"})


async def reset_password(token: str, request: Request, response: Response) -> JSONResponse:
    """Reset password using the token from the email link.

    Mirrors: resetPassword in authControllers.js
    PUT /api/v1/password/reset/{token}
    """
    hashed = hashlib.sha256(token.encode()).hexdigest()

    user = await User.find_one(
        User.reset_password_token == hashed,
        User.reset_password_expire > datetime.now(timezone.utc),
    )
    if not user:
        raise ErrorHandler("Password reset token is invalid or has expired", 400)

    data = await request.json()
    new_password = data.get("password")
    confirm_password = data.get("confirmPassword")

    if new_password != confirm_password:
        raise ErrorHandler("Passwords do not match", 400)

    user.password = new_password
    user.reset_password_token = None
    user.reset_password_expire = None
    await user.save()

    token_data = send_token(user, 200, response)
    return JSONResponse({"success": True, **token_data})


# ── Current User ───────────────────────────────────────────────────────────────

async def get_current_user_details(current_user: User) -> JSONResponse:
    """Return the authenticated user's profile.

    Mirrors: getCurrentUserDetails in authControllers.js
    GET /api/v1/me
    """
    return JSONResponse({"success": True, "user": _serialize_user(current_user)})


async def change_password(request: Request, current_user: User, response: Response) -> JSONResponse:
    """Change the current user's password.

    Mirrors: changePassword in authControllers.js
    PUT /api/v1/password/update
    """
    data = await request.json()
    old_password = data.get("oldPassword")
    new_password = data.get("password")
    confirm_password = data.get("confirmPassword")

    if not current_user.match_password(old_password):
        raise ErrorHandler("Old password is incorrect", 400)

    if new_password != confirm_password:
        raise ErrorHandler("Passwords do not match", 400)

    current_user.password = new_password
    await current_user.save()

    token_data = send_token(current_user, 200, response)
    return JSONResponse({"success": True, **token_data})


async def update_user_profile(request: Request, current_user: User) -> JSONResponse:
    """Update name and email of the current user.

    Mirrors: updateUserProfile in authControllers.js
    PUT /api/v1/me/updateprofile
    """
    data = await request.json()
    name = data.get("name", current_user.name)
    email = data.get("email", current_user.email)

    await current_user.set({"name": name, "email": email, "updated_at": datetime.now(timezone.utc)})
    user = await User.get(str(current_user.id))
    return JSONResponse({"success": True, "user": _serialize_user(user)})


async def upload_avatar(request: Request, current_user: User) -> JSONResponse:
    """Placeholder for avatar upload — real implementation requires cloud storage.

    Mirrors: uploadAvatar in authControllers.js
    PUT /api/v1/me/upload_avatar
    """
    data = await request.json()
    avatar_url = data.get("avatar")

    await current_user.set({
        "avatar": {"public_id": "avatar_placeholder", "url": avatar_url},
        "updated_at": datetime.now(timezone.utc),
    })
    user = await User.get(str(current_user.id))
    return JSONResponse({"success": True, "user": _serialize_user(user)})


# ── Admin User Management ─────────────────────────────────────────────────────

async def get_all_admin_users() -> JSONResponse:
    """Return all users (admin only).

    Mirrors: getAllAdminUsers in authControllers.js
    GET /api/v1/admin/users
    """
    users = await User.find_all().to_list()
    return JSONResponse({"success": True, "users": [_serialize_user(u) for u in users]})


async def get_admin_user_by_id(user_id: str) -> JSONResponse:
    """Return a single user by ID (admin only).

    Mirrors: getAdminUserById in authControllers.js
    GET /api/v1/admin/userid/{id}
    """
    user = await _get_user_or_404(user_id)
    return JSONResponse({"success": True, "user": _serialize_user(user)})


async def update_user_profile_by_admin(user_id: str, request: Request) -> JSONResponse:
    """Update name, email, and role of any user (admin only).

    Mirrors: updateUserProfileByAdmin in authControllers.js
    PUT /api/v1/admin/user/update/{id}
    """
    user = await _get_user_or_404(user_id)
    data = await request.json()

    update = {k: v for k, v in data.items() if k in ("name", "email", "role")}
    update["updated_at"] = datetime.now(timezone.utc)
    await user.set(update)

    user = await User.get(user_id)
    return JSONResponse({"success": True, "user": _serialize_user(user)})


async def delete_user_by_admin(user_id: str) -> JSONResponse:
    """Delete a user account (admin only).

    Mirrors: deleteUserByAdmin in authControllers.js
    DELETE /api/v1/admin/user/delete/{id}
    """
    user = await _get_user_or_404(user_id)
    await user.delete()
    return JSONResponse({"success": True, "message": "User deleted"})


# ── Helpers ───────────────────────────────────────────────────────────────────

async def _get_user_or_404(user_id: str) -> User:
    try:
        user = await User.get(user_id)
    except Exception:
        user = None
    if not user:
        raise ErrorHandler(f"User not found with id: {user_id}", 404)
    return user


def _serialize_user(user: User) -> dict:
    """Convert a User document to a JSON-serialisable dict (no password)."""
    d = user.model_dump()
    d["_id"] = str(user.id)
    d.pop("id", None)
    d.pop("password", None)
    d.pop("reset_password_token", None)
    d.pop("reset_password_expire", None)
    for key in ("created_at", "updated_at"):
        if isinstance(d.get(key), datetime):
            d[key] = d[key].isoformat()
    return d
